src/main/java/chess/dao/ChessGameDao.java

```diff
@@ -6,44 +6,70 @@
 import chess.domain.position.File;
 import chess.domain.position.Position;
 import chess.domain.position.Rank;
+import chess.domain.state.State;
 import chess.dto.ChessGameDto;
+import chess.exception.ExistGameException;
 import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import org.springframework.jdbc.core.JdbcTemplate;
 import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public class ChessGameDao {
 
-    private JdbcTemplate jdbcTemplate;
+    private final JdbcTemplate jdbcTemplate;
 
     public ChessGameDao(JdbcTemplate jdbcTemplate) {
         this.jdbcTemplate = jdbcTemplate;
     }
 
-    public void save(ChessGameDto chessGameDto) {
-        String sql = "insert into chessgame (game_name, turn) values (?, ?)";
-        jdbcTemplate.update(sql, chessGameDto.getGameName(), chessGameDto.getTurn());
+    public Long save(ChessGameDto chessGameDto, String password) {
+        validateDuplicateGameName(chessGameDto.getGameName());
+
+        String sql = "insert into chessgame (game_name, turn, password) values (?, ?, ?)";
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        jdbcTemplate.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(sql, new String[]{"id"});
+            ps.setString(1, chessGameDto.getGameName());
+            ps.setString(2, chessGameDto.getTurn());
+            ps.setString(3, password);
+            return ps;
+        }, keyHolder);
+
+        return Objects.requireNonNull(keyHolder.getKey()).longValue();
     }
 
-    public void update(ChessGameDto chessGameDto) {
-        String sql = "update chessgame set turn = ? where game_name = ?";
-        jdbcTemplate.update(sql, chessGameDto.getTurn(), chessGameDto.getGameName());
+    private void validateDuplicateGameName(String gameName) {
+        String sql = "select count(*) from CHESSGAME where game_name = ?";
```

Comment:
```suggestion
    private void validateDuplicateGameName(String gameName) {
        String sql = "select * from CHESSGAME where game_name = ? LIMIT 1";
```
limit 1을 사용하면 모든 테이블을 탐색하지 않을 수도 있겠네요

Reply:
LIMIT을 사용하여 하나의 row가 검색되면 더 이상 탐색하지 않음으로써 모든 테이블의 row를 탐색하지 않아도 되어 성능개선이 있을 수 있을 것 같습니다! (특히 row가 10,000개 이상 100,000개 이상과 같이 정말 많이지는 경우 그 이점을 더욱 명확히 느낄 수 있을 것 같습니다!) 개선해보도록 하겠습니다😃

Reply:
jdbcTemplate.queryForObject()는 조회 결과가 없는 경우 `EmptyResultDataAccessException` 을 발생시킵니다.
따라서 위의 코멘트 변경과 함께 해당 쿼리문을 수정하여 다음과 같이 코드를 작성할 수 있었습니다.

```java
private void validateDuplicateGameName(String gameName) {
    try {
        chessGameDao.findOneByGameName(gameName);
        throw new IllegalGameProgressException("동일한 이름의 게임이 이미 존재합니다.");
    } catch (EmptyResultDataAccessException ignored) {
    }
}
```

즉, 우리가 원하는 실행흐름(정상흐름)이 `EmptyResulstDataAccessException`이기 때문에 위와 같이 코드를 작성하게 되었습니다. 그런데 try-catch구문에서 오히려 예외가 발생할 때 정상흐름이 된다는게 어색하게 느껴졌습니다.
1. 실무에서도 위와같은 코드가 생길 때가 종종 있는지 궁금합니다!
2. 위와 같은 코드에서 발생할 수 있는 문제점이 있을까요? 만약 있다면 어떤 것이 있는지 궁금합니다!
저는 우선 가독성 측면에서 단점이 있다고 생각했습니다. (이 코드를 처음보는 사람이 이해하기 위해서는 많이 시간이 들어가지 않을까?)

Reply:
디우 이 부분은 DM으로 답변드렸습니다! exists 쿼리를 이용해볼 수 있을 것 같아요

Reply:
감사합니다!!!😃